src/main/kotlin/org/rust/ide/hints/type/RsChainMethodTypeHintsProvider.kt

```diff
@@ -69,40 +74,48 @@ class RsChainMethodTypeHintsProvider : InlayHintsProvider<RsChainMethodTypeHints
 
     override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
         object : FactoryInlayHintsCollector(editor) {
-
-            val typeHintsFactory = RsTypeHintsPresentationFactory(
-                if (settings.iteratorSpecialCase) file as? RsElement else null,
-                factory,
-                true
-            )
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, true)
 
             override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
                 if (DumbService.isDumb(element.project)) return true
                 if (element !is RsMethodCall) return true
                 if (!element.isLastInChain) return true
                 if (!element.isEnabledByCfg) return true
 
+                val (lookup, items) = element.implLookupAndKnownItems
+                val iterator = items.Iterator?.let { BoundElement(it) }
+
                 val chain = collectChain(element)
                 var lastType: Ty? = null
                 for (call in chain.dropLast(1)) {
-                    if (call.type != TyUnknown && call.isLastOnLine) {
-                        if (settings.showSameConsecutiveTypes || call.type != lastType) {
-                            presentTypeForMethodCall(call)
+                    val type = normalizeType(call.type, lookup, iterator)
+                    if (type != TyUnknown && call.isLastOnLine) {
+                        if (settings.showSameConsecutiveTypes || type != lastType) {
+                            presentTypeForMethodCall(call, type)
                         }
-                        lastType = call.type
+                        lastType = type
                     }
                 }
 
                 return true
             }
 
-            private fun presentTypeForMethodCall(call: RsMethodCall) {
+            private fun presentTypeForMethodCall(call: RsMethodCall, type: Ty) {
                 val project = call.project
-                val type = call.type
                 val presentation = typeHintsFactory.typeHint(type)
                 val finalPresentation = presentation.withDisableAction(project)
                 sink.addInlineElement(call.endOffset, true, finalPresentation)
             }
+
+            private fun normalizeType(type: Ty, lookup: ImplLookup, iteratorTrait: BoundElement<RsTraitItem>?): Ty {
```

Comment:
Let's add a comment about what's happening here, like this:
/** Returns fake `impl Iterator<Item=...>` type if [type] implements `Iterator` trait */

Also, looks like you missed to check `settings.iteratorSpecialCase` somewhere